src/core/mount.c

```diff
@@ -142,6 +152,104 @@ static bool mount_needs_quota(const MountParameters *p) {
                                  "usrquota\0" "grpquota\0" "quota\0" "usrjquota\0" "grpjquota\0");
 }
 
+static int alloc_mount_hash(void) {
+        Hashmap *by_mnt_id;
+        Hashmap *by_device;
+
+        by_mnt_id = hashmap_new(NULL);
+        if (!by_mnt_id)
+                return -ENOMEM;
+        by_device = hashmap_new(NULL);
+        if (!by_device) {
+                hashmap_free(by_mnt_id);
+                return -ENOMEM;
+        }
+        mount_hash.by_mnt_id = by_mnt_id;
+        mount_hash.by_device = by_device;
+
+        return 0;
+}
+
+static int mount_hash_put(Mount *m, int mnt_id, const char *what) {
+        Mount *mnt;
+        int dev_ref;
+        int r;
+
+        assert(m);
+
+        if (!mount_hash.by_mnt_id) {
+                if (alloc_mount_hash() != 0)
+                        return -ENOMEM;
+        }
+
+        if (mnt_id == -1)
+                return 0;
+
+        mnt = hashmap_get(mount_hash.by_mnt_id, INT_TO_PTR(mnt_id));
+        if (!mnt) {
+                r = hashmap_put(mount_hash.by_mnt_id, INT_TO_PTR(mnt_id), m);
+                if (r < 0)
+                        return -ENOMEM;
+        }
+
+        dev_ref = PTR_TO_INT(hashmap_get(mount_hash.by_device, what));
```

Comment:
hmm, what's the valid range for a mount id? 0…INT_MAX? or 1…INT_MAX? If the former, we should do this differently, since hashmap_get() returns NULL if entry does not exist

Reply:
Ha, a good question, one that I've talked about a bit upstream too.

At the time I wrote this the definition was 0..INT_MAX but now I'm told that 0 isn't used (I didn't spot that in the kernel implementation) but I'm also told that using -1 for an invalid mount id will continue to work in the face of possible changes.

I'll need to verify these things when the changes are merged into the upstream kernel and revisit this then.
Ideally, from my perspective, 0 as an invalid mount id would be best but I'll probably need to stick with -1.